File: frontend/src/api/ocr.ts
import api from './config';
import { Task } from './tasks';

export interface OcrPageRef {
  collection: string;
  registre: string;
  page: string;
}

export interface OcrScopeItem {
  collection: string;
  registre?: string;
}

// Filtre d'état appliqué à un périmètre par le backend, miroir de `OcrService._SCOPE_ONLY`.
export type OcrScopeOnly = 'all' | 'missing' | 'done';

export const ocrApi = {
  // Enfile une tâche OCR (suivie via le système de tâches générique).
  //
  // Deux façons de désigner les pages, cumulables : `pages` les nomme une à une, `scopes`
  // désigne des registres entiers que le backend développe depuis le disque. Le périmètre
  // évite d'envoyer les centaines de milliers de noms d'une grosse collection — et surtout
  // de les fabriquer depuis la pagination, ce que le client ne sait pas faire (la largeur du
  // champ numérique est perdue dans le motif).
  //
  // `timeout: 0` : le backend liste chaque registre concerné avant d'enfiler — sur un partage
  // réseau et une sélection de centaines de registres, cela dépasse le filet de 60 s.
  run: async (
    seg_model_id: string,
    ocr_model_id: string,
    pages: OcrPageRef[],
    scopes?: OcrScopeItem[],
    scope_only: OcrScopeOnly = 'all',
  ): Promise<Task> => {
    const response = await api.post(
      '/api/ocr/run', { seg_model_id, ocr_model_id, pages, scopes, scope_only }, { timeout: 0 });
    return response.data;
  },

  // Stems (nom sans extension) des pages déjà transcrites pour (collection, registre, modèle).
  donePages: async (collection: string, registre: string, model: string): Promise<string[]> => {
    const response = await api.get('/api/ocr/done', { params: { collection, registre, model } });
    return response.data.stems;
  },

  // Pages sans transcription pour un modèle (périmètre optionnel ; absent = tout).
  missingPages: async (model: string, scope?: OcrScopeItem[]): Promise<OcrPageRef[]> => {
    const response = await api.post('/api/ocr/missing', { model, scope });
    return response.data.pages;
  },
};

File: backend/routers/ocr.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List, Optional

from ocr_service import OcrService, NoPagesToProcess
from task_service import TaskConflict, TaskUnavailable

router = APIRouter()


class OcrPageRef(BaseModel):
    collection: str
    registre: str
    page: str


class OcrScopeItem(BaseModel):
    collection: str
    registre: Optional[str] = None


# Filtres d'état d'un périmètre, miroir de `OcrService._SCOPE_ONLY`.
SCOPE_ONLY = ('all', 'missing', 'done')


class OcrRunRequest(BaseModel):
    seg_model_id: str
    ocr_model_id: str
    # Pages nommées une à une, et/ou registres entiers développés par le backend depuis le
    # disque. Le périmètre évite au client d'envoyer — et de fabriquer — les centaines de
    # milliers de noms d'une grosse collection.
    pages: List[OcrPageRef] = []
    scopes: Optional[List[OcrScopeItem]] = None
    scope_only: str = 'all'


class OcrMissingRequest(BaseModel):
    model: str
    scope: Optional[List[OcrScopeItem]] = None


@router.post("/run")
def run_ocr(request: OcrRunRequest):
    """Ajoute une tâche OCR à la file (gérée par le moteur de tâches générique)."""
    if not request.pages and not request.scopes:
        raise HTTPException(status_code=400, detail="Aucune page à traiter.")
    if not request.seg_model_id or not request.ocr_model_id:
        raise HTTPException(status_code=400, detail="Modèle de segmentation et modèle OCR requis.")
    if request.scope_only not in SCOPE_ONLY:
        raise HTTPException(status_code=400, detail=f"Filtre de périmètre inconnu : {request.scope_only}")
    pages = [p.model_dump() for p in request.pages]
    scopes = [s.model_dump() for s in request.scopes] if request.scopes else None
    try:
        return OcrService.enqueue(request.seg_model_id, request.ocr_model_id, pages,
                                  scopes=scopes, scope_only=request.scope_only)
    except NoPagesToProcess as e:
        raise HTTPException(status_code=400, detail=str(e))
    except TaskConflict as e:
        raise HTTPException(status_code=409, detail=str(e))
    except TaskUnavailable as e:
        raise HTTPException(status_code=503, detail=str(e))


@router.get("/done")
def ocr_done(collection: str, registre: str, model: str):
    """Stems des pages déjà transcrites pour (collection, registre, modèle)."""
    return {"stems": OcrService.done_stems(collection, registre, model)}


@router.post("/missing")
def ocr_missing(request: OcrMissingRequest):
    """Pages sans transcription pour un modèle (périmètre optionnel)."""
    if not request.model:
        raise HTTPException(status_code=400, detail="Modèle OCR requis.")
    scope = [s.model_dump() for s in request.scope] if request.scope else None
    pages = OcrService.missing_pages(request.model, scope)
    return {"model": request.model, "count": len(pages), "pages": pages}
